src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -14861,6 +14923,594 @@ public void Main(string key)
             c.VerifyDiagnostics();
         }
 
+        [Fact]
+        [WorkItem(29295, "https://github.com/dotnet/roslyn/issues/29295")]
+        public void CatchException()
+        {
+            var c = CreateCompilation(@"
+class C
+{
+    void M()
+    {
+        try
+        {
+            System.Console.WriteLine();
+        }
+        catch (System.Exception e)
+        {
+            e.ToString();
+        }
+    }
+}
+", options: WithNonNullTypesTrue());
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(29295, "https://github.com/dotnet/roslyn/issues/29295")]
+        public void CatchException_WithWhenIsOperator()
+        {
+            var c = CreateCompilation(@"
+class C
+{
+    void M()
+    {
+        try
+        {
+            System.Console.WriteLine();
+        }
+        catch (System.Exception e) when (!(e is System.ArgumentException))
+        {
+            e.ToString();
+        }
+    }
+}
+", options: WithNonNullTypesTrue());
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(26739, "https://github.com/dotnet/roslyn/issues/26739")]
+        public void RefParameter_TopLevelNullability()
+        {
+            var c = CreateCompilation(@"
+class C
+{
+    public void M()
+    {
+        string? s = string.Empty;
+        M(ref s);
+        s.ToString(); // 1
+
+        string s2 = string.Empty;
+        M(ref s2); // 2
+        s2.ToString(); // 3
+    }
+    void M(ref string? s) => throw null!;
+}
+", options: WithNonNullTypesTrue());
+
+            c.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         s.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(8, 9),
+                // (11,15): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         M(ref s2); // 2
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "s2").WithLocation(11, 15),
+                // (12,9): warning CS8602: Possible dereference of a null reference.
+                //         s2.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(12, 9)
+                );
+        }
+
+        [Fact]
+        [WorkItem(26739, "https://github.com/dotnet/roslyn/issues/26739")]
+        public void RefParameter_TopLevelNullability_Suppressed()
+        {
+            var c = CreateCompilation(@"
+class C
+{
+    public void M()
+    {
+        string? s = string.Empty;
+        M(ref s!);
+        s.ToString(); // no warning
```

Comment:
Not sure I agree with this one not getting a warning? Unlike the case below, the `!` on the `ref` isn't suppressing any warnings. It feels wrong that putting a `!` where there isn't any warning changes top-level nullability.

Reply:
The definition of the suppression operator has two aspects: (1) it changes the top-level nullability, and (2) it suppresses warnings. Presently, the former is unconditional and isn't dependent on the latter.
[https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-reference-types.md#suppression-operator](https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-reference-types.md#suppression-operator)-

In the case of a `ref` argument, we could discuss whether the `!` should change the top-level nullability on the way in, or on the way out. But we already discussed that for `out` arguments with LDM (ie. the suppression does affect the value on the way out), so it seems logical to do the same for `ref`.

We can discuss with Chuck/Neal at next opportunity, or add an open issue to list [https://github.com/dotnet/csharplang/issues/2201](https://github.com/dotnet/csharplang/issues/2201)

---
In reply to: [258177260](https://github.com/dotnet/roslyn/pull/33447#discussion_r258177260) [](ancestors = 258177260)

Reply:
I added a question to the linked issue.

---
In reply to: [258237238](https://github.com/dotnet/roslyn/pull/33447#discussion_r258237238) [](ancestors = 258237238,258177260)